File: spawnllm/spec.py
"""Spec-driven run configuration shared across every backend."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from pydantic import BaseModel

    from spawnllm.types import ProviderName, TSettingSource


@dataclass(frozen=True, slots=True)
class AppleConfig:
    """Apple Foundation Models knobs applied only by the Apple backend.

    `use_case` and `guardrails` select the `SystemLanguageModelUseCase` and
    `SystemLanguageModelGuardrails` the `spawnllm-apple` sidecar builds its
    session with. The sampling knobs are flat rather than a nested mode because
    Apple exposes `SamplingMode` as a factory (`greedy()` / `random()`) that no
    serializable value can carry: `sampling` picks the factory and `sampling_top`,
    `sampling_probability_threshold`, and `sampling_seed` are the arguments
    `random` takes. `None` everywhere leaves the framework default.

    Example:
        >>> AppleConfig(use_case="content_tagging", sampling="random", sampling_top=20)

    Raises:
        ValueError: When a `random`-only argument is set without `sampling="random"`,
            a combination the framework would silently discard; when `sampling_top`
            and `sampling_probability_threshold` are set together, a pair the
            framework rejects; or when `sampling_seed` is negative, which the
            framework's `UInt64` cannot carry.
    """

    use_case: Literal["general", "content_tagging"] = "general"
    guardrails: Literal["default", "permissive_content_transformations"] = "default"
    instructions: str | None = None
    temperature: float | None = None
    maximum_response_tokens: int | None = None
    sampling: Literal["greedy", "random"] | None = None
    sampling_top: int | None = None
    sampling_probability_threshold: float | None = None
    sampling_seed: int | None = None

    def __post_init__(self) -> None:
        if self.sampling != "random" and any(
            knob is not None for knob in (self.sampling_top, self.sampling_probability_threshold, self.sampling_seed)
        ):
            raise ValueError(
                "AppleConfig sampling_top, sampling_probability_threshold, and sampling_seed require sampling='random'"
            )
        if self.sampling_top is not None and self.sampling_probability_threshold is not None:
            raise ValueError("AppleConfig accepts either sampling_top or sampling_probability_threshold, not both")
        if self.sampling_seed is not None and self.sampling_seed < 0:
            raise ValueError("AppleConfig sampling_seed is unsigned; the framework takes a UInt64")


@dataclass(frozen=True, slots=True)
class ClaudeConfig:
    """Claude CLI flag passthrough applied only by the Claude backend.

    Fields map one-to-one onto `claude` flags: the agent/system-prompt knobs
    (`permission_mode`, `mcp_config`, `strict_mcp`, `append_system_prompt`,
    `system_prompt`, `settings`, `disallowed_tools`) and orthogonal extras
    (`max_turns`, `max_budget_usd`, `tools`, `disable_slash_commands`,
    `output_format`, `verbose`). `tools` selects the built-in toolset: `None`
    keeps the CLI default, `()` disables every built-in tool, and names
    restrict the session to those tools. `setting_sources` selects
    `--setting-sources`: `None` drops the host's user settings (`("project",)`,
    or nothing when `isolated`), so a run never inherits the caller's hooks.

    Example:
        >>> ClaudeConfig(permission_mode="bypassPermissions", strict_mcp=True)
        >>> ClaudeConfig(tools=())  # bare session: no built-in tools
        >>> ClaudeConfig(setting_sources=("user", "project", "local"))  # CLI default
    """

    permission_mode: str | None = None
    mcp_config: str | None = None
    strict_mcp: bool = False
    setting_sources: tuple[TSettingSource, ...] | None = None
    append_system_prompt: str | None = None
    system_prompt: str | None = None
    settings: str | None = None
    disallowed_tools: tuple[str, ...] = ()
    max_turns: int | None = None
    max_budget_usd: float | None = None
    tools: tuple[str, ...] | None = None
    disable_slash_commands: bool = False
    output_format: str | None = None
    verbose: bool = False


@dataclass(frozen=True, slots=True)
class CodexConfig:
    """Codex CLI knobs applied only by the Codex backend.

    `service_tier` (default `"fast"`) is emitted as `-c service_tier=<value>` on
    every invocation, isolated or not: isolated runs pass `--ignore-user-config`,
    which drops a `service_tier` pin in the user's `~/.codex/config.toml`, and the
    standard tier turns long prompts into multi-minute runs. Set it to `None` to
    drop the flag; an isolated run still passes `--ignore-user-config`, so a
    user-level tier pin applies only with `isolated=False`. `developer_instructions` injects
    the system-prompt layer via `-c developer_instructions=<value>`, serialized as
    a TOML string so any text — multi-line, or TOML-ambiguous words like `true` —
    arrives as a string.

    Example:
        >>> CodexConfig(sandbox="read-only", enable_mcp=True)
    """

    sandbox: str | None = None
    enable_hooks: bool = False
    enable_mcp: bool = False
    service_tier: str | None = "fast"
    developer_instructions: str | None = None


@dataclass(frozen=True, slots=True)
class GeminiConfig:
    """Gemini CLI knobs applied only by the Gemini and Antigravity backends.

    Example:
        >>> GeminiConfig(approval_mode="auto", extensions=("search",))
    """

    approval_mode: str | None = None
    extensions: tuple[str, ...] | None = None


type ProviderConfig = AppleConfig | ClaudeConfig | CodexConfig | GeminiConfig


@dataclass(frozen=True, slots=True)
class RunSpec:
    """A single configured run, translated per backend at execution time.

    Common fields are interpreted by every backend; `provider_configs` carries
    optional per-provider flag passthrough that only the matching backend reads.
    `model` is a literal provider model id (`opus`, `sonnet`, …) passed straight
    through with no tier mapping. `isolated` (default `True`) runs the backend
    against a fresh, host-free config home so a spawned CLI ignores ambient
    settings, MCP servers, and hooks. `api_auth` (default `False`) strips the
    provider's API-key environment variables from the child process so the CLI
    bills the logged-in subscription; `True` inherits the environment untouched.
    Structured output comes from either a `response_model` (validated to a model)
    or a raw `schema` (a JSON-Schema dict or pre-serialized string, passed to the
    provider verbatim, with nothing to validate); setting both raises `ValueError`.

    Example:
        >>> RunSpec(prompt="ping", model="opus")
    """

    prompt: str
    model: str
    response_model: type[BaseModel] | None = None
    schema: dict[str, object] | str | None = None
    agent: bool = False
    isolated: bool = True
    cwd: str | None = None
    env: dict[str, str] | None = None
    api_auth: bool = False
    timeout: int = 180
    max_attempts: int = 5
    provider_configs: dict[ProviderName, ProviderConfig] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.response_model is not None and self.schema is not None:
            raise ValueError("RunSpec accepts either response_model or schema, not both")

    def config_for[T: ProviderConfig](self, kind: type[T]) -> T | None:
        """Return the first provider config that is an instance of `kind`, or None."""
        return next((c for c in self.provider_configs.values() if isinstance(c, kind)), None)

File: go/spec.go
package spawnllm

import (
	"encoding/json"
	"time"
)

// ModelTier is an abstract model size resolved to a provider model id via core
// capabilities. The zero value is Small.
type ModelTier string

// The abstract model sizes each backend maps to a concrete model id.
const (
	Small  ModelTier = "small"
	Medium ModelTier = "medium"
	Large  ModelTier = "large"
)

func (t ModelTier) orDefault() ModelTier {
	if t == "" {
		return Small
	}
	return t
}

// AppleUseCase names the SystemLanguageModelUseCase the sidecar builds its
// session with. The zero value is AppleUseCaseGeneral.
type AppleUseCase string

// The use cases the on-device model is specialized for.
const (
	AppleUseCaseGeneral        AppleUseCase = "general"
	AppleUseCaseContentTagging AppleUseCase = "content_tagging"
)

func (u AppleUseCase) orDefault() AppleUseCase {
	if u == "" {
		return AppleUseCaseGeneral
	}
	return u
}

// AppleGuardrails names the SystemLanguageModelGuardrails the sidecar builds its
// session with. The zero value is AppleGuardrailsDefault.
type AppleGuardrails string

// The guardrail settings the on-device model runs under.
const (
	AppleGuardrailsDefault                          AppleGuardrails = "default"
	AppleGuardrailsPermissiveContentTransformations AppleGuardrails = "permissive_content_transformations"
)

func (g AppleGuardrails) orDefault() AppleGuardrails {
	if g == "" {
		return AppleGuardrailsDefault
	}
	return g
}

// AppleSampling picks the SamplingMode factory the sidecar calls. The empty
// string leaves the framework default.
type AppleSampling string

// The sampling modes the on-device model decodes with.
const (
	AppleSamplingGreedy AppleSampling = "greedy"
	AppleSamplingRandom AppleSampling = "random"
)

// AppleConfig passes on-device Foundation Models knobs through to the
// spawnllm-apple sidecar. Empty string and zero fields are unset; SamplingTop,
// SamplingProbabilityThreshold, and SamplingSeed apply only to
// AppleSamplingRandom, and Top and ProbabilityThreshold are mutually exclusive —
// Run and RunOn reject either combination rather than dropping the knob.
type AppleConfig struct {
	Instructions                 string
	UseCase                      AppleUseCase
	Guardrails                   AppleGuardrails
	Temperature                  float64
	MaximumResponseTokens        int
	Sampling                     AppleSampling
	SamplingTop                  int
	SamplingProbabilityThreshold float64
	SamplingSeed                 uint64
}

// ClaudeConfig passes flags through to the claude CLI. Empty string and zero
// fields are unset; Tools distinguishes nil (CLI default) from an empty slice
// (disable every built-in tool).
type ClaudeConfig struct {
	PermissionMode       string
	MCPConfig            string
	StrictMCP            bool
	AppendSystemPrompt   string
	SystemPrompt         string
	Settings             string
	DisallowedTools      []string
	MaxTurns             int
	MaxBudgetUSD         float64
	Tools                []string
	DisableSlashCommands bool
	OutputFormat         string
	Verbose              bool
}

// CodexConfig passes knobs through to the codex CLI.
type CodexConfig struct {
	Sandbox     string
	EnableHooks bool
	EnableMCP   bool
	// ServiceTier nil pins "fast"; a pointer to "" drops the flag, and any
	// other value selects that tier.
	ServiceTier           *string
	DeveloperInstructions string
}

// GeminiConfig passes knobs through to the gemini and antigravity CLIs.
// Extensions distinguishes nil (default set) from an empty slice (no extensions).
type GeminiConfig struct {
	ApprovalMode string
	Extensions   []string
}

// ProviderConfigs carries the per-provider flag passthrough a RunSpec applies;
// only the matching backend's config is read.
type ProviderConfigs struct {
	Apple  *AppleConfig
	Claude *ClaudeConfig
	Codex  *CodexConfig
	Gemini *GeminiConfig
}

// RunSpec is a single configured run, translated per backend at execution time.
// Model is a literal provider model id. UseHostConfig false (the zero value) runs
// against a fresh, host-free config home; Timeout is per-attempt (0 → 180s) and
// MaxAttempts bounds the transient-retry loop (0 → 5).
type RunSpec struct {
	Prompt string
	Model  string
	// Schema, when set, must encode a JSON object and is passed to the provider verbatim.
	Schema        json.RawMessage
	Agent         bool
	UseHostConfig bool
	APIAuth       bool
	Dir           string
	Env           map[string]string
	Timeout       time.Duration
	MaxAttempts   int
	Providers     ProviderConfigs
}

func (s RunSpec) timeout() time.Duration {
	if s.Timeout <= 0 {
		return 180 * time.Second
	}
	return s.Timeout
}

func (s RunSpec) maxAttempts() int {
	if s.MaxAttempts <= 0 {
		return 5
	}
	return s.MaxAttempts
}

// coreSpec is the RunSpec in the core wire shape: every optional maps to a
// pointer so an unset field serializes to null, and the config sub-structs carry
// every field the core requires present.
type coreSpec struct {
	Prompt         string          `json:"prompt"`
	Model          string          `json:"model"`
	Agent          bool            `json:"agent"`
	Isolated       bool            `json:"isolated"`
	APIAuth        bool            `json:"api_auth"`
	Timeout        int64           `json:"timeout"`
	MaxAttempts    int64           `json:"max_attempts"`
	Schema         json.RawMessage `json:"schema"`
	Apple          *coreApple      `json:"apple"`
	Claude         *coreClaude     `json:"claude"`
	Codex          *coreCodex      `json:"codex"`
	Gemini         *coreGemini     `json:"gemini"`
	OpenAIEndpoint *coreOpenAI     `json:"openai_endpoint"`
}

type coreApple struct {
	Instructions                 *string         `json:"instructions"`
	UseCase                      AppleUseCase    `json:"use_case"`
	Guardrails                   AppleGuardrails `json:"guardrails"`
	Temperature                  *float64        `json:"temperature"`
	MaximumResponseTokens        *int            `json:"maximum_response_tokens"`
	Sampling                     *AppleSampling  `json:"sampling"`
	SamplingTop                  *int            `json:"sampling_top"`
	SamplingProbabilityThreshold *float64        `json:"sampling_probability_threshold"`
	SamplingSeed                 *uint64         `json:"sampling_seed"`
}

type coreClaude struct {
	AppendSystemPrompt   *string  `json:"append_system_prompt"`
	DisableSlashCommands bool     `json:"disable_slash_commands"`
	DisallowedTools      []string `json:"disallowed_tools"`
	MaxBudgetUSD         *float64 `json:"max_budget_usd"`
	MaxTurns             *int     `json:"max_turns"`
	MCPConfig            *string  `json:"mcp_config"`
	OutputFormat         *string  `json:"output_format"`
	PermissionMode       *string  `json:"permission_mode"`
	Settings             *string  `json:"settings"`
	StrictMCP            bool     `json:"strict_mcp"`
	SystemPrompt         *string  `json:"system_prompt"`
	Tools                []string `json:"tools"`
	Verbose              bool     `json:"verbose"`
}

type coreCodex struct {
	DeveloperInstructions *string `json:"developer_instructions"`
	EnableHooks           bool    `json:"enable_hooks"`
	EnableMCP             bool    `json:"enable_mcp"`
	Sandbox               *string `json:"sandbox"`
	ServiceTier           *string `json:"service_tier"`
}

type coreGemini struct {
	ApprovalMode *string  `json:"approval_mode"`
	Extensions   []string `json:"extensions"`
}

type coreOpenAI struct {
	APIKey  string `json:"api_key"`
	BaseURL string `json:"base_url"`
	Model   string `json:"model"`
}

func optString(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func optInt(n int) *int {
	if n == 0 {
		return nil
	}
	return &n
}

func optUint64(n uint64) *uint64 {
	if n == 0 {
		return nil
	}
	return &n
}

func optFloat(f float64) *float64 {
	if f == 0 {
		return nil
	}
	return &f
}

func (s RunSpec) core() coreSpec {
	schema := s.Schema
	if len(schema) == 0 {
		schema = nil
	}
	return coreSpec{
		Prompt:      s.Prompt,
		Model:       s.Model,
		Agent:       s.Agent,
		Isolated:    !s.UseHostConfig,
		APIAuth:     s.APIAuth,
		Timeout:     int64(s.timeout() / time.Second),
		MaxAttempts: int64(s.maxAttempts()),
		Schema:      schema,
		Apple:       coreAppleOf(s.Providers.Apple),
		Claude:      coreClaudeOf(s.Providers.Claude),
		Codex:       coreCodexOf(s.Providers.Codex),
		Gemini:      coreGeminiOf(s.Providers.Gemini),
	}
}

func coreAppleOf(c *AppleConfig) *coreApple {
	if c == nil {
		return nil
	}
	var sampling *AppleSampling
	if c.Sampling != "" {
		sampling = &c.Sampling
	}
	return &coreApple{
		Instructions:                 optString(c.Instructions),
		UseCase:                      c.UseCase.orDefault(),
		Guardrails:                   c.Guardrails.orDefault(),
		Temperature:                  optFloat(c.Temperature),
		MaximumResponseTokens:        optInt(c.MaximumResponseTokens),
		Sampling:                     sampling,
		SamplingTop:                  optInt(c.SamplingTop),
		SamplingProbabilityThreshold: optFloat(c.SamplingProbabilityThreshold),
		SamplingSeed:                 optUint64(c.SamplingSeed),
	}
}

func coreClaudeOf(c *ClaudeConfig) *coreClaude {
	if c == nil {
		return nil
	}
	return &coreClaude{
		AppendSystemPrompt:   optString(c.AppendSystemPrompt),
		DisableSlashCommands: c.DisableSlashCommands,
		DisallowedTools:      c.DisallowedTools,
		MaxBudgetUSD:         optFloat(c.MaxBudgetUSD),
		MaxTurns:             optInt(c.MaxTurns),
		MCPConfig:            optString(c.MCPConfig),
		OutputFormat:         optString(c.OutputFormat),
		PermissionMode:       optString(c.PermissionMode),
		Settings:             optString(c.Settings),
		StrictMCP:            c.StrictMCP,
		SystemPrompt:         optString(c.SystemPrompt),
		Tools:                c.Tools,
		Verbose:              c.Verbose,
	}
}

func coreCodexOf(c *CodexConfig) *coreCodex {
	if c == nil {
		return nil
	}
	tier := c.ServiceTier
	if tier == nil {
		fast := "fast"
		tier = &fast
	} else if *tier == "" {
		tier = nil
	}
	return &coreCodex{
		DeveloperInstructions: optString(c.DeveloperInstructions),
		EnableHooks:           c.EnableHooks,
		EnableMCP:             c.EnableMCP,
		Sandbox:               optString(c.Sandbox),
		ServiceTier:           tier,
	}
}

func coreGeminiOf(c *GeminiConfig) *coreGemini {
	if c == nil {
		return nil
	}
	return &coreGemini{
		ApprovalMode: optString(c.ApprovalMode),
		Extensions:   c.Extensions,
	}
}
